### PR TITLE
Proper Handling of the URL Configuration Change

### PR DESCRIPTION
**Pull request by ngondalia**
This commit provides proper handling of url configuration change. This fixes the issue of when the user changes the URL of the server, the user would need to reopen vscode so that all of the manangers can be initialized with the proper tspClientURL. With this commit, all tspClient dependent components are notified of a URL change.

Signed-off-by: Neel Gondalia ngondalia@blackberry.com

### CODE REVIEW

**Comment by williamsyang-work** [comment]
Hey Neel,

I currently have a PR under review that addresses this exact problem as well.  https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/196.

Maybe you can take a look and see if this works for you?

**Review by bhufmann — COMMENTED** [review]
I was testing this change and the URL change is propagated correctly. However, I noticed that when change the URL, the webviews don't populate and I see the error message below in the console log. In my test I changed the URL from `localhost` to `10.0.0.122`. All the updates are done but, the html of the webview is not resolve again and it has the old URL for the `connect-src`. For example, the method `resolveWebviewView()` which calls `_getHtmlForWebview` of `TraceExplorerOpenedTracesViewProvider` is only called once and hence the `connect-src` is not updated. I'm not sure how to fix that. Please note it's a general problem and not related to this change. Any ideas how to fix it?

image

**Comment by ngondalia** [comment]
> I was testing this change and the URL change is propagated correctly. However, I noticed that when change the URL, the webviews don't populate and I see the error message below in the console log. In my test I changed the URL from `localhost` to `10.0.0.122`. All the updates are done but, the html of the webview is not resolve again and it has the old URL for the `connect-src`. For example, the method `resolveWebviewView()` which calls `_getHtmlForWebview` of `TraceExplorerOpenedTracesViewProvider` is only called once and hence the `connect-src` is not updated. I'm not sure how to fix that. Please note it's a general problem and not related to this change. Any ideas how to fix it?
> 
> image

I pushed a patch for it! Can you please re-test it and let me know? 

From the docs of webview.html property:
"
HTML contents of the webview.

This should be a complete, valid html document. Changing this property causes the webview to be reloaded.
"

Based on this, you can just assign the updated html content to the webview and it will get reloaded.

**Comment by bhufmann** [comment]
> > I was testing this change and the URL change is propagated correctly. However, I noticed that when change the URL, the webviews don't populate and I see the error message below in the console log. In my test I changed the URL from `localhost` to `10.0.0.122`. All the updates are done but, the html of the webview is not resolve again and it has the old URL for the `connect-src`. For example, the method `resolveWebviewView()` which calls `_getHtmlForWebview` of `TraceExplorerOpenedTracesViewProvider` is only called once and hence the `connect-src` is not updated. I'm not sure how to fix that. Please note it's a general problem and not related to this change. Any ideas how to fix it?
> > image
> 
> I pushed a patch for it! Can you please re-test it and let me know?
> 
> From the docs of webview.html property: " HTML contents of the webview. This should be a complete, valid html document. Changing this property causes the webview to be reloaded. "
> 
> Based on this, you can just assign the updated html content to the webview and it will get reloaded.

Yes, this did the trick. I tested it and works. Thanks.

I'm going to do a review now. Once this PR merged, @williamsyang-work will base his work on it.

File: vscode-trace-extension/src/trace-explorer/available-views/trace-explorer-available-views-webview-provider.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/available-views/trace-explorer-available-views-webview-provider.ts:39]: I think you can call the lines below in method `updateTraceServerUrl()`. With this you can avoid adding this public API.

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-webview-provider.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-webview-provider.ts:62]: I think you can call the lines below in method `updateTraceServerUrl()`. With this you can avoid adding this public API.

File: vscode-trace-extension/src/trace-explorer/properties/trace-explorer-properties-view-webview-provider.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/properties/trace-explorer-properties-view-webview-provider.ts:15]: I think you can call the lines below in method `updateTraceServerUrl()`. With this you can avoid adding this public API.

File: vscode-trace-extension/src/trace-explorer/time-range/trace-explorer-time-range-data-webview-provider.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/time-range/trace-explorer-time-range-data-webview-provider.ts:23]: I think you can call the lines below in method `updateTraceServerUrl()`. With this you can avoid adding this public API.

File: vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts
Comment (bhufmann) [inline vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts:74]: I think you can call the lines below in method `updateTraceServerUrl()`. With this you can avoid adding this public API.

**Review by bhufmann — CHANGES_REQUESTED** [review]

**Review by bhufmann — APPROVED** [review]
Looks good to me. Thanks a lot for this contribution.

**Comment by ngondalia** [comment]
> Looks good to me. Thanks a lot for this contribution. Since you're committer, you can submit it. Please use `Rebase and Merge`.

I don't have that option. It seems like i still don't have write access.

**Comment by williamsyang-work** [comment]
> > I was testing this change and the URL change is propagated correctly. However, I noticed that when change the URL, the webviews don't populate and I see the error message below in the console log. In my test I changed the URL from `localhost` to `10.0.0.122`. All the updates are done but, the html of the webview is not resolve again and it has the old URL for the `connect-src`. For example, the method `resolveWebviewView()` which calls `_getHtmlForWebview` of `TraceExplorerOpenedTracesViewProvider` is only called once and hence the `connect-src` is not updated. I'm not sure how to fix that. Please note it's a general problem and not related to this change. Any ideas how to fix it?
> > image
> 
> I pushed a patch for it! Can you please re-test it and let me know?
> 
> From the docs of webview.html property: " HTML contents of the webview. This should be a complete, valid html document. Changing this property causes the webview to be reloaded. "
> 
> Based on this, you can just assign the updated html content to the webview and it will get reloaded.

Awesome find.  This bug has been on my radar for a while, I appreciate the fix!

**Review by hriday-panchasara — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
